PEP 698: Fix typos and suggest edits (#2990)
python/peps: pep-0698.rst

Runtime Override Checks in Python
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Today, there is an |overrides-library|_ that provides decorators
> ``@overrides`` (sic) and ``@final`` and will enforce them at runtime.

Reviewer: ```suggestion
Today, there is an `Overrides library <https://pypi.org/project/overrides/>`__
that provides decorators ``@overrides`` [sic] and ``@final``
and will enforce them at runtime.
```

At least IMO, it's major overkill and a rather ugly and complicated hack to use two layers of substitutions here, just all to get ``overrides`` to display as a literal. There's really no practical need to put the proper name of a project in literals, unless it is being referred to in an actual literal context like a CLI command, Python code, string value, etc; otherwise, it's just visual noise, IMO, that dilutes the semantic meaning of literals in the source and the output. And nor does the project itself use this convention in it's own [Readme](https://github.com/mkorpela/overrides) (where it uses "overrides") or [Releases page](https://github.com/mkorpela/overrides/releases)_ (where it uses "Overrides").

Author: I have no objections to reverting this -- the project name was formatted as code-literal elsewhere in the PEP, and that's my personal preference; thus the suggestion.

(I completely agree that the double substitution is bulky, and I was expecting pushback on those grounds. But, AFAIK it's the only way to achieve this in reST.)

If we revert here, then I'd advocate for converting formatting of *all* references to the project to whatever form, 'overrides' or 'Overrides'.

Reviewer: Ah, I see—I should have checked that, sorry.

> If we revert here, then I'd advocate for converting formatting of all references to the project to whatever form, 'overrides' or 'Overrides'.

That would be my suggestion as well, as personally I'm not really a big fan of formatting project names in literals everywhere without a clear semantically-appropriate reason, as it just adds noise and complexity in both the source and the output, and leads to cases like this. Though, even if not, I'd still say it's not worth all this just for the link, viz. "A foolish consistency..." However, it's ultimately up to @stroxler as PEP author.

Reviewer 2: I think we could drop the formatting of the library name everywhere, and just format the decorator names. The library name itself isn't really a code snippet.

Runtime Override Checks in Python
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Today, there is an `Overrides library <https://pypi.org/project/overrides/>`__

Reviewer 2: What do the double-underscores do?

It seems like single underscore works for inline links, as far as I can tell from (the rst quickref)[https://docutils.sourceforge.io/docs/user/rst/quickref.html] double-underscores are used for anonymous hyperlinks, which I think is not relevant here?

But I'll be honest RST confuses me so I may be wrong

Author: Great question. Reaching back to my first learning of reST, I had thought they were necessary for inline, direct links to places.

Now that I read the hyperlink docs more closely -- I don't think they do what I thought they did. Single underscores work fine, and I think are more correct. I'll put those back. 😬

Reviewer: It makes the link anonymous, which means you can't later reference the same link target later in the document. This is useful if you have a link like `` `here <https://example.com>`__ ``. Conversely, it is not what you want if you do want to later reference the same link target, e.g.

```rst
Today, there is an `Overrides library <https://example.com>`_ ...

The `Overrides library` is a reference implementation ...
```

This is exactly the situation here, and it was a silly mistake for me to use it here, since in my other suggestions I deliberately reference the link target multiple other places.

As described above, adding ``__override__`` may fail at runtime, in which
case we will simply return the argument as-is.

> In addition, even in cases where it does work it may be difficult for users to

Reviewer: ```suggestion
In addition, even in cases where it does work, it may be difficult for users to
```

Nitpick: Needs a comma here to separate these two clauses

Author: Mm, I saw this but left it thinking it was a stylistic choice.

Reviewer: Same for my first pass as I've become a lot more conservative about suggesting others add commas in PEPs lately, but it is more than merely stylistic in this case.

case we will simply return the argument as-is.

In addition, even in cases where it does work it may be difficult for users

> to correctly work with multiple decorators, because successfully setting
> the ``__override__`` field on the final output requires understanding

Reviewer: ```suggestion
to correctly work with multiple decorators, because successfully ensuring
the ``__override__`` attribute is set on the final output requires understanding
```

* Tweak the wording to more accurately reflect the apparent intent here is ensuring the attribute still exists on the final chained output, as opposed to directly setting it there
* Isn't this an attribute, not a "field" (whatever that is?)

Author: @stroxler, what's your opinion on the semantics here?

Reviewer 2: I think @CAM-Gerlach's wording is better (although the initial proposal was better than the original, which was not very clear at all).

We considered including a class decorator ``@require_explicit_overrides``, which
would have provided a way for base classes to declare that all subclasses must

> use the ``@override`` decorator on method overrides. The ``overrides`` library has a

Reviewer: ```suggestion
use the ``@override`` decorator on method overrides. The `Overrides library`_ has a
```

As discussed above, instead of making it a literal (for some reason), why not just link it instead, which is substantially more appropriate and useful to the reader?

  be considerable benefits.
- We believe that it would be rarely used and catch relatively few bugs.

  - The author of the

>     `Overrides package <https://pypi.org/project/overrides/>`_ has

Reviewer: Here, you _could_ reuse the target from above, but it runs the risk of breaking the link if the other reference is removed or changed, so best to just keep it as is, I think.

Author: :+1:

Author: If we don't plan to re-use the reference name elsewhere, would it be better to make it anonymous so as to avoid cluttering the namespace?

How are name collisions handled by Sphinx?

It might not matter as much for the PEPs repo; but, could it lead to incorrect references in a larger docset?

Reviewer: > If we don't plan to re-use the reference name elsewhere, would it be better to make it anonymous so as to avoid cluttering the namespace?

The way I view it, if the URL is more or less pseudo-canonical for the link text in the context (i.e., we would be unlikely to link `Overrides package` to anything else in this PEP), then there's no harm in omitting the `_` just in case.

In any case, given this is perhaps one of the final significant edits to an accepted PEP, this is a bit of just counting angels on pinheads. :)

> How are name collisions handled by Sphinx?

Technically speaking, hyperlink targets are a feature of baseline reST and processed by Docutils, not Sphinx. Thus, they are namespaced per-document, and accordingly do not depend on the number of documents present (which would actually be a much bigger problem here than most places, otherwise).

Defining or attempting to use a duplicate hyperlink target produces loud warnings, and will cause the build to exit if `-W` is passed (which we're working toward by default on this repo, and already do on CPython, Devguide, etc), or if the docutils errorlevel is set (as we do here).

For example, if a document has:

```rst
`test link <https://example.com>`_

`test link`_

`test link <https://example.org>`_

`test link`_
```

the build will output:

```
/Path/To/FIle.rst:2: WARNING: Duplicate explicit target name: "test link".
/Path/To/File.rst:13: ERROR: Duplicate target name, cannot be used as a unique reference: "test link".
/Path/To/File.rst:17: ERROR: Duplicate target name, cannot be used as a unique reference: "test link".
```

Author: Cool, thanks very much for the explainer!

>     `noted <https://discuss.python.org/t/pep-698-a-typing-override-decorator/20839/4>`_

Reviewer: ```suggestion
    `noted <https://discuss.python.org/t/pep-698-a-typing-override-decorator/20839/4>`__
```

This would be a spot where an anonymous hyperlink would be appropriate, as `noted` is an ambiguous and easy to re-use link target name.

This document is placed in the public domain or under the
CC0-1.0-Universal license, whichever is more permissive.

> 
> .. |overrides-library| replace:: ``overrides`` library
> 
> .. _overrides-library: https://pypi.org/project/overrides/

Reviewer: ```suggestion
```

As discussed above, all this extra complexity really doesn't seem necessary or prudent.
